### PR TITLE
Tag list

### PR DESCRIPTION
**Pull request by hbjorgo**
Added new endpoint that searches the tag api and returns all items found with with an additional property that states if each tag is in the preservation scope.

Divided the tag controller in two - one that deals with all preservation tags and one that deals with tag api tags. They have separete routes: "tags/available" and "tags/preserved"

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQuery.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Collections.Generic;
+using MediatR;
+using ServiceResult;
+
+namespace Equinor.Procosys.Preservation.Query.AllAvailableTagsQuery
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQuery.cs]: TagSearchQuery (all namespaces)

File: src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQuery.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public class GetAllAvailableTagsQuery : IRequest<Result<List<ProcosysTagDto>>>
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQuery.cs]: Rename to SearchTagsQuery

File: src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQuery.cs (hunk 1 above)
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQuery.cs]: rename namespace .. to be discussed

File: src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+using System.Threading.Tasks;
+using Equinor.Procosys.Preservation.Domain;
+using Equinor.Procosys.Preservation.Domain.AggregateModels.TagAggregate;
+using Equinor.Procosys.Preservation.MainApi.Tag;
+using MediatR;
+using ServiceResult;
+
+namespace Equinor.Procosys.Preservation.Query.AllAvailableTagsQuery
+{
+    public class GetAllAvailableTagsQueryHandler : IRequestHandler<GetAllAvailableTagsQuery, Result<List<ProcosysTagDto>>>
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs]: SearchTagsQueryHandler

File: src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            _tagRepository = tagRepository;
+            _tagApiService = tagApiService;
+            _plantProvider = plantProvider;
+        }
+
+        public async Task<Result<List<ProcosysTagDto>>> Handle(GetAllAvailableTagsQuery request, CancellationToken cancellationToken)
+        {
+            var apiTags = await _tagApiService.GetTags(_plantProvider.Plant, request.ProjectName, request.StartsWithTagNo) ?? new List<ProcosysTagOverview>();
+            var presTags = await _tagRepository.GetAllAsync() ?? new List<Tag>();
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs]: I think we should get all preserved tags within project

File: src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        private readonly IPlantProvider _plantProvider;
+
+        public GetAllAvailableTagsQueryHandler(ITagRepository tagRepository, ITagApiService tagApiService, IPlantProvider plantProvider)
+        {
+            _tagRepository = tagRepository;
+            _tagApiService = tagApiService;
+            _plantProvider = plantProvider;
+        }
+
+        public async Task<Result<List<ProcosysTagDto>>> Handle(GetAllAvailableTagsQuery request, CancellationToken cancellationToken)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs]: format this method. Many long lines

File: src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            var apiTags = await _tagApiService.GetTags(_plantProvider.Plant, request.ProjectName, request.StartsWithTagNo) ?? new List<ProcosysTagOverview>();
+            var presTags = await _tagRepository.GetAllAsync() ?? new List<Tag>();
+
+            // Join all tags from API with preservation tags on TagNo. If a tag is not in preservation scope, use default value (null).
+            var combinedTags = apiTags
+                .GroupJoin(presTags, apiTag => apiTag.TagNo, presTag => presTag.TagNo, (x, y) => new { ApiTag = x, PresTag = y })
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs]: Impressing code, but I don't have a chance to understand it :-|

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Tags/AvailableTagsController.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+using Microsoft.AspNetCore.Mvc;
+using ServiceResult.ApiExtensions;
+
+namespace Equinor.Procosys.Preservation.WebApi.Controllers.Tags
+{
+    /// <summary>
+    /// Handles requests that deal with all ProCoSys tags, not just preservation
+    /// </summary>
+    [ApiController]
+    [Route("Tags/Available")]
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/Tags/AvailableTagsController.cs]: Search

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Tags/AvailableTagsController.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    public class AvailableTagsController : ControllerBase
+    {
+        private readonly IMediator _mediator;
+
+        public AvailableTagsController(IMediator mediator) => _mediator = mediator;
+
+        /// <summary>
+        /// Gets all tags from ProCoSys and encriches them with preservation data
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/Tags/AvailableTagsController.cs]: Search

File: src/Equinor.Procosys.Preservation.WebApi/Controllers/Tags/AvailableTagsController.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        /// </summary>
+        /// <param name="projectName"></param>
+        /// <param name="startsWithTagNo"></param>
+        /// <returns>All ProCoSys tags that match the search parameters</returns>
+        [HttpGet]
+        public async Task<ActionResult<List<ProcosysTagDto>>> GetAllAvailableTags([FromQuery] string projectName, [FromQuery] string startsWithTagNo)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Controllers/Tags/AvailableTagsController.cs]: SearchTags

File: src/tests/Equinor.Procosys.Preservation.Query.Tests/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandlerTests.cs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        [TestMethod]
+        public async Task Handle_ReturnsOkResult()
+        {
+            _tagApiServiceMock
+                .Setup(x => x.GetTags("PCS$TESTPLANT", "ProjectName", "TagNo"))
+                .Returns(Task.FromResult(_apiTags));
+
+            var dut = new GetAllAvailableTagsQueryHandler(_tagRepositoryMock.Object, _tagApiServiceMock.Object, _plantProviderMock.Object);
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Query.Tests/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandlerTests.cs]: make _dut private in class and setup in Setup()

File: src/tests/Equinor.Procosys.Preservation.Query.Tests/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandlerTests.cs [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+            var query = new GetAllAvailableTagsQuery("ProjectName", "TagNo");
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Query.Tests/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandlerTests.cs]: make _query private in class and setup in Setup()

File: src/tests/Equinor.Procosys.Preservation.WebApi.Tests/Controllers/Tags/AvailableTagsControllerTests.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        private readonly List<ProcosysTagDto> _listWithTwoItems = new List<ProcosysTagDto>
+        {
+            new ProcosysTagDto("TagNo1", "Desc1", "PO1", "CommPkg1", "McPkg1", true),
+            new ProcosysTagDto("TagNo2", "Desc2", "PO2", "CommPkg2", "McPkg2", false)
+        };
+
+        [TestInitialize]
+        public void Setup()
+        {
+            _dut = new AvailableTagsController(_mediatorMock.Object);
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.WebApi.Tests/Controllers/Tags/AvailableTagsControllerTests.cs]: expr.body

File: src/tests/Equinor.Procosys.Preservation.WebApi.Tests/Controllers/Tags/AvailableTagsControllerTests.cs [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            _mediatorMock
+                .Setup(x => x.Send(It.IsAny<GetAllAvailableTagsQuery>(), It.IsAny<CancellationToken>()))
+                .Returns(Task.FromResult(new SuccessResult<List<ProcosysTagDto>>(_listWithTwoItems) as Result<List<ProcosysTagDto>>));
+
+            var result = await _dut.GetAllAvailableTags("ProjectName", "TagNo");
+
+            Assert.IsInstanceOfType(result.Result, typeof(OkObjectResult));
+
+            Assert.AreEqual(2, (((OkObjectResult)result.Result).Value as List<ProcosysTagDto>).Count);
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.WebApi.Tests/Controllers/Tags/AvailableTagsControllerTests.cs]: Resharper complains: Possible nullpointer exception (same at line 83). Refactor?

Comment (hbjorgo) [reply src/tests/Equinor.Procosys.Preservation.WebApi.Tests/Controllers/Tags/AvailableTagsControllerTests.cs]: Added more asserts.

File: src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs (hunk 6 above)
Comment (hbjorgo) [reply src/Equinor.Procosys.Preservation.Query/AllAvailableTagsQuery/GetAllAvailableTagsQueryHandler.cs]: Won't fix :P

File: src/tests/Equinor.Procosys.Preservation.Query.Tests/TagApiQueries/SearchTags/GetAllAvailableTagsQueryHandlerTests.cs [hunk 14]
```diff
@@ -82,10 +81,9 @@ public async Task Handle_ReturnsOkResult()
                 .Setup(x => x.GetTags("PCS$TESTPLANT", "ProjectName", "TagNo"))
                 .Returns(Task.FromResult(_apiTags));
 
-            var dut = new GetAllAvailableTagsQueryHandler(_tagRepositoryMock.Object, _tagApiServiceMock.Object, _plantProviderMock.Object);
-            var query = new GetAllAvailableTagsQuery("ProjectName", "TagNo");
+            var query = new SearchTagsQuery("ProjectName", "TagNo");
```
Comment (janid1967) [inline src/tests/Equinor.Procosys.Preservation.Query.Tests/TagApiQueries/SearchTags/GetAllAvailableTagsQueryHandlerTests.cs]: make private _query and Setup once

**Pull request merged**